test/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+    }
+}
+
+/// Return a list with all possible Linux test cases for the given CWE.
+pub fn linux_test_cases(cwe: &'static str, check_name: &'static str) -> Vec<CweTestCase> {
+    new_test_cases(cwe, ARCHITECTURES, COMPILERS, check_name)
+        .into_iter()
+        .filter(|test| test.architecture != "ppc" || test.compiler != "clang")
+        .collect()
+}
+
+/// Return a list with all possible Windows test cases for the given CWE
+pub fn windows_test_cases(cwe: &'static str, check_name: &'static str) -> Vec<CweTestCase> {
+    new_test_cases(cwe, WINDOWS_ARCHITECTURES, WINDOWS_COMPILERS, check_name)
+}
+
+/// Generate test cases for all combinations of CPU architecture and compiler given.
+pub fn new_test_cases(
+    cwe: &'static str,
+    architectures: &[&'static str],
+    compilers: &[&'static str],
+    check_name: &'static str,
+) -> Vec<CweTestCase> {
+    let mut vec = Vec::new();
+    for architecture in architectures {
+        for compiler in compilers {
+            vec.push(CweTestCase {
+                cwe,
+                architecture,
+                compiler,
+                check_name,
+                skipped: false,
+            });
+        }
+    }
+    vec
+}
+
+/// Return a list of all possible test cases (Linux and Windows) for the given CWE.
+pub fn all_test_cases(cwe: &'static str, check_name: &'static str) -> Vec<CweTestCase> {
+    let mut vec = linux_test_cases(cwe, check_name);
+    vec.append(&mut windows_test_cases(cwe, check_name));
+    vec
+}
+
+/// Print the error messages of failed checks.
+/// The `error_log` tuples are of the form `(check_filename, error_message)`.
+pub fn print_errors(error_log: Vec<(String, String)>) {
+    for (filepath, error) in error_log {
+        println!("{}", format!("+++ Error for {} +++", filepath).red());
+        println!("{}", error);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[ignore]
+    fn cwe_415() {
+        let mut error_log = Vec::new();
+        let mut tests = all_test_cases("cwe_415", "Memory");
+
+        mark_architecture_skipped(&mut tests, "mips64"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "mips64el"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "mips"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "mipsel"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "ppc64"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "ppc64le"); // TODO: Check reason for failure!
+        mark_architecture_skipped(&mut tests, "ppc"); // TODO: Check reason for failure!
+        mark_skipped(&mut tests, "x86", "gcc"); // TODO: Check reason for failure!
+        mark_compiler_skipped(&mut tests, "mingw32-gcc"); // TODO: Check reason for failure!
+
+        for test_case in tests {
+            if let Err(error) = test_case.run_test("[CWE415]", 2) {
```

Comment:
Do you think it would be sensible to assigned the number to a variable like `let expected_occurrences: u8 = 2;`? It says it at the function header of run_test but I have forgotten what the number stood for when I came across the call and had to check the header again.

Reply:
Sounds like a good idea. Implemented.